dotnet/src/Connectors/Connectors.Memory.Kusto/KustoMemoryStore.cs

```diff
@@ -0,0 +1,411 @@
+        {
+            var key = reader.GetString(0);
+            var metadata = reader.GetString(1);
+            DateTime? timestamp = reader.IsDBNull(2) ? null : reader.GetDateTime(2);
+            var _embedding = withEmbeddings ? reader.GetString(3) : default;
+
+            var kustoRecord = new KustoMemoryRecord(key, metadata, _embedding, timestamp);
+
+            yield return kustoRecord.ToMemoryRecord();
+        }
+    }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<string> GetCollectionsAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        var result = await this._adminClient
+            .ExecuteControlCommandAsync<TablesShowCommandResult>(
+                this._database,
+                CslCommandGenerator.GenerateTablesShowCommand() + $" | where TableName startswith '{c_collectionPrefix}'",
+                GetClientRequestProperties()
+            ).ConfigureAwait(false);
+
+        foreach (var item in result)
+        {
+            yield return GetCollectionName(item.TableName);
+        }
+    }
+
+    /// <inheritdoc/>
+    public async Task<(MemoryRecord, double)?> GetNearestMatchAsync(string collectionName, Embedding<float> embedding, double minRelevanceScore = 0, bool withEmbedding = false, CancellationToken cancellationToken = default)
+    {
+        var result = this.GetNearestMatchesAsync(collectionName, embedding, 1, minRelevanceScore, withEmbedding, cancellationToken);
+        return await result.FirstOrDefaultAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<(MemoryRecord, double)> GetNearestMatchesAsync(string collectionName, Embedding<float> embedding, int limit, double minRelevanceScore = 0, bool withEmbeddings = false, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        var similarityQuery = $"{this.GetBaseQuery(collectionName)} | extend similarity=series_cosine_similarity_fl('{KustoSerializer.SerializeEmbedding(embedding)}', {s_collectionColumns[2].Name}, 1, 1)";
+
+        if (minRelevanceScore != 0)
+        {
+            similarityQuery += $" | where similarity > {minRelevanceScore}";
+        }
+
+        similarityQuery += $" | top {limit} by similarity desc";
+        // reorder to make it easier to ignore the embedding (key, metadata, timestamp, similarity, embedding
+        // Using tostring to make it easier to parse the result. There are probably better ways we should explore
+        similarityQuery += "| project " +
+            // Key
+            $"{s_collectionColumns[0].Name}, " +
+            // Metadata
+            $"{s_collectionColumns[1].Name}=tostring({s_collectionColumns[1].Name}), " +
+            // Timestamp
+            $"{s_collectionColumns[3].Name}, " +
+            "similarity, " +
+            // Embedding
+            $"{s_collectionColumns[2].Name}=tostring({s_collectionColumns[2].Name})";
```

Comment:
The queries are not identical. Carving out common code for potential future re-use doesn't seem like a good investment IMHO. If further changes will be needed in the future, I can definitely do that, but I feel like that's somewhat unlikely.

Reply:
Got it, I see `similarity` property there. Keeping them in both places will also work at the moment. As soon as we add similar query a third time, most probably it will go to separate method, following the "Rule of Three" :)